README.md

[![Travis build status](https://travis-ci.com/martakarass/runstats.svg?branch=master)](https://travis-ci.com/martakarass/runstats) [![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/martakarass/runstats?branch=master&svg=true)](https://ci.appveyor.com/project/martakarass/runstats) [![Coverage status](https://codecov.io/gh/martakarass/runstats/branch/master/graph/badge.svg)](https://codecov.io/github/martakarass/runstats?branch=master)

<!-- README.md is generated from README.Rmd. Please edit that file -->
runstats
========

Package `runstats` provides methods for fast computation of running sample statistics for time series. The methods utilize Convolution Theorem to compute convolutions via Fast Fourier Transform (FFT). Implemented running statistics include:

1.  mean,
2.  standard deviation,
3.  variance,
4.  covariance,
5.  correlation,
6.  euclidean distance.

### Website

Package website is located [here](https://martakarass.github.io/runstats/).

### Installation

``` r
devtools::install_github("martakarass/runstats")
```

### Usage

``` r
library(runstats)

## Example: running correlation
x0 <- sin(seq(0, 2 * pi * 5, length.out = 1000))
x  <- x0 + rnorm(1000, sd = 0.1)
pattern <- x0[1:100]
out1 <- RunningCor(x, pattern)
out2 <- RunningCor(x, pattern, circular = TRUE)

## Example: running mean
x <- cumsum(rnorm(1000))
out1 <- RunningMean(x, W = 100)
out2 <- RunningMean(x, W = 100, circular = TRUE)
```

### Running statistics

To better explain the details of running statistics, package's function `runstats.demo(func.name)` allows to visualize how the output of each running statistics method is generated. To run the demo, use `func.name` being one of the methods' names:

1.  `"RunningMean"`,
2.  `"RunningSd"`,
3.  `"RunningVar"`,
4.  `"RunningCov"`,
5.  `"RunningCor"`,
6.  `"RunningL2Norm"`.

``` r
## Example: demo for running correlation method  
runstats.demo("RunningCor")
```

![](https://i.imgur.com/JQObAu7.gif)

``` r
## Example: demo for running mean method 
runstats.demo("RunningMean")
```

![](https://i.imgur.com/FWah3aG.gif)

### Performance

We use `rbenchmark` to measure elapsed time of `RunningCov` execution, for different lengths of time-series `x` and fixed length of the shorter pattern `y`.

``` r
library(rbenchmark)

set.seed (20181010)
x.N.seq <- 10^(3:7)
x.list  <- lapply(x.N.seq, function(N) runif(N))
y <- runif(100)

## Benchmark execution time of RunningCov 
out.df <- data.frame()
for (x.tmp in x.list){
  out.df.tmp <- benchmark("runstats" = runstats::RunningCov(x.tmp, y),
                          replications = 10,
                          columns = c("test", "replications", "elapsed",
                                      "relative", "user.self", "sys.self"))
  out.df.tmp$x_length <- length(x.tmp)
  out.df.tmp$pattern_length <- length(y)
  out.df <- rbind(out.df, out.df.tmp)
}
```

``` r
knitr::kable(out.df)
```

| test     |  replications|  elapsed|  relative|  user.self|  sys.self|  x\_length|  pattern\_length|
|:---------|-------------:|--------:|---------:|----------:|---------:|----------:|----------------:|
| runstats |            10|    0.003|         1|      0.003|     0.000|       1000|              100|
| runstats |            10|    0.031|         1|      0.025|     0.006|      10000|              100|
| runstats |            10|    0.385|         1|      0.328|     0.056|     100000|              100|
| runstats |            10|    5.694|         1|      5.509|     0.171|    1000000|              100|
| runstats |            10|  126.393|         1|    121.007|     4.954|   10000000|              100|

##### Compare with a conventional method

To compare `RunStats` performance with "conventional" loop-based way of computing running covariance in `R`, we use `rbenchmark` package to measure elapsed time of `RunStats::RunningCov` and running covariance implemented with `sapply` loop, for different lengths of time-series `x` and fixed length of the shorter time-series `y`.

``` r
## Conventional approach 
RunningCov.sapply <- function(x, y){
  l_x <- length(x)
  l_y <- length(y)
  sapply(1:(l_x - l_y + 1), function(i){
    cov(x[i:(i+l_y-1)], y)
  })
}

set.seed (20181010)

out.df2 <- data.frame()
for (x.tmp in x.list[c(1,2,3,4)]){
  out.df.tmp <- benchmark("conventional" = RunningCov.sapply(x.tmp, y),
                          "runstats" = runstats::RunningCov(x.tmp, y),
                          replications = 10,
                          columns = c("test", "replications", "elapsed",
                                      "relative", "user.self", "sys.self"))
  out.df.tmp$x_length <- length(x.tmp)
  out.df2 <- rbind(out.df2, out.df.tmp)
}
```

Benchmark results

``` r
library(ggplot2)

plt1 <- 
  ggplot(out.df2, aes(x = x_length, y = elapsed, color = test)) + 
  geom_line() + geom_point(size = 3) + scale_x_log10() + 
  theme_minimal(base_size = 14) + 
  labs(x = "Vector length of x",
       y = "Elapsed [s]", color = "Method", 
       title = "Running covariance rbenchmark") + 
  theme(legend.position = "bottom")
plt2 <- 
  plt1 + 
  scale_y_log10() + 
  labs(y = "Log of elapsed [s]")

cowplot::plot_grid(plt1, plt2, nrow = 1, labels = c('A', 'B'))
```

<img src="man/figures/README-unnamed-chunk-13-1.png" width="100%" />

Platform information

``` r
sessioninfo::platform_info()
#>  setting  value                       
#>  version  R version 3.5.2 (2018-12-20)
#>  os       macOS Mojave 10.14.2        
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2019-02-28
```


NEWS.md
# runstats 1.0.1

* Added a `NEWS.md` file to track changes to the package.


tests/testthat/test-RunningSd.R
context("Checking RunningSd")

test_that("Checking mean of RunningSd hasn't changed  (with circular = TRUE)" , {
  set.seed(20180618)
  x <- rnorm(1000)
  output <- RunningSd(x, 100, circular = TRUE)
  expect_equal(mean(output), 0.990468893492985)
})

test_that("Checking mean of RunningSd hasn't changed  (with circular = FALSE)", {
  set.seed(20180618)
  x <- rnorm(1000)
  output <- RunningSd(x, 100, circular = FALSE)
  expect_equal(mean(output, na.rm = TRUE), 0.985677717702135)
})



RunningSd.CONV <- function(x, W, circular){

  out <- numeric()
  l_x <- length(x)

  out <- sapply(1:(l_x - W + 1), function(i){
    segm <- x[i:(i+W-1)]
    sd(segm)
  })

  if (circular) {
    out.tail <- sapply((l_x - W + 2):l_x, function(i){
      segm <- c(x[i:l_x], x[1:(W-l_x+i-1)])
      sd(segm)
    })
  } else {
    out.tail <- rep(NA, W-1)
  }

  out <- c(out, out.tail)

  return(out)
}


test_that("Checking RunningSd via FFT returns what is equivalent with conventional computation (with circular = TRUE)", {
  set.seed(20180618)
  x <- rnorm(1000)
  output      <- RunningSd(x, 100, circular = TRUE)
  output.conv <- RunningSd.CONV(x, 100, circular = TRUE)
  max.diff    <- max(abs(output - output.conv))
  expect_equal(max.diff, 0, tolerance = 1e-8)
})


test_that("Checking RunningSd via FFT returns what is equivalent with conventional computation (with circular = FALSE)", {

  set.seed(20180618)

  x <- rnorm(1000)
  output      <- RunningSd(x, 100, circular = FALSE)
  output.conv <- RunningSd.CONV(x, 100, circular = FALSE)

  expect_true(all(is.na(output) == is.na(output.conv)))

  max.diff  <- max(abs(output - output.conv), na.rm = TRUE)
  expect_equal(max.diff, 0, tolerance = 1e-8)

})



tests/testthat/test-RunningCor.R
context("Checking RunningCor")


RunningCor.CONV <- function(x, y, circular){

  out <- numeric()
  W <- length(y)
  l_x <- length(x)

  out <- sapply(1:(l_x - W + 1), function(i){
    segm <- x[i:(i+W-1)]
    cor(segm, y)
  })

  if (circular) {
    out.tail <- sapply((l_x - W + 2):l_x, function(i){
      segm <- c(x[i:l_x], x[1:(W-l_x+i-1)])
      cor(segm, y)
    })
  } else {
    out.tail <- rep(NA, W-1)
  }

  out <- c(out, out.tail)

  return(out)
}


test_that("Checking RunningCor via FFT returns what is equivalent with conventional computation (with circular = TRUE)", {
  x <- sin(seq(0, 1, length.out = 1000) * 2 * pi * 6)
  y <- x[1:100]
  output      <- RunningCor(x, y, circular = TRUE)
  output.conv <- RunningCor.CONV(x, y, circular = TRUE)
  max.diff    <- max(abs(output - output.conv))
  expect_equal(max.diff, 0, tolerance = 1e-8)
})


test_that("Checking RunningCor via FFT returns what is equivalent with conventional computation (with circular = FALSE)", {

  x <- sin(seq(0, 1, length.out = 1000) * 2 * pi * 6)
  y <- x[1:100]

  output      <- RunningCor(x, y, circular = FALSE)
  output.conv <- RunningCor.CONV(x, y, circular = FALSE)

  max.diff    <- max(abs(output - output.conv), na.rm = TRUE)
  expect_equal(max.diff, 0, tolerance = 1e-8)

  expect_true(all(is.na(output) == is.na(output.conv)))
})



tests/testthat/test-RunningCov.R
context("Checking RunningCov")


RunningCov.CONV <- function(x, y, circular){

  out <- numeric()
  W <- length(y)
  l_x <- length(x)

  out <- sapply(1:(l_x - W + 1), function(i){
    segm <- x[i:(i+W-1)]
    cov(segm, y)
  })

  if (circular) {
    out.tail <- sapply((l_x - W + 2):l_x, function(i){
      segm <- c(x[i:l_x], x[1:(W-l_x+i-1)])
      cov(segm, y)
    })
  } else {
    out.tail <- rep(NA, W-1)
  }

  out <- c(out, out.tail)

  return(out)
}


test_that("Checking RunningCov via FFT returns what is equivalent with conventional computation (with circular = TRUE)", {
  x <- sin(seq(0, 1, length.out = 1000) * 2 * pi * 6)
  y <- x[1:100]
  output      <- RunningCov(x, y, circular = TRUE)
  output.conv <- RunningCov.CONV(x, y, circular = TRUE)
  max.diff    <- max(abs(output - output.conv))
  expect_equal(max.diff, 0, tolerance = 1e-8)
})


test_that("Checking RunningCov via FFT returns what is equivalent with conventional computation (with circular = FALSE)", {

  x <- sin(seq(0, 1, length.out = 1000) * 2 * pi * 6)
  y <- x[1:100]

  output      <- RunningCov(x, y, circular = FALSE)
  output.conv <- RunningCov.CONV(x, y, circular = FALSE)

  expect_true(all(is.na(output) == is.na(output.conv)))

  max.diff    <- max(abs(output - output.conv), na.rm = TRUE)
  expect_equal(max.diff, 0, tolerance = 1e-8)

  expect_true(all(is.na(output) == is.na(output.conv)))

})



R/run_demo.R


#' Generates plots for demo of package functions which take time series and
#' window width parameters
#'
#' @param func runstats package core function
#' @param plt.title.vec vector of function-specific plot titles
#'
#' @importFrom grDevices rgb
#' @importFrom graphics abline lines par plot points polygon title
#'
#' @return \code{NULL}
#'
#' @examples
#' \dontrun{
#' func <- RunningMean
#' vec <- c("black: x\nred: W-width running window",
#'          "RunningMean(x, W)",
#'          "RunningMean(x, W, circular = TRUE)")
#' plot.no.pattern(func, vec)
#' }
#'
#' @noRd
#'
plot.no.pattern <- function(func, plt.title.vec, func.name){

  ## Fixed constraints
  N <- 600
  W <- 100
  by.val <- 10

  ## Simulate signal
  set.seed(123)
  if (func.name %in% c("RunningSd", "RunningVar")){
    rnorm.vec <- rnorm(N, mean = 0, sd = c(seq(1, 100, length.out = N/2), rev(seq(1, 100, length.out = N/2))))
    x <- rnorm.vec
  } else {
    rnorm.vec <- rnorm(N)
    x <- cumsum(rnorm.vec)
  }

  ## Function output
  out.cir.F <- do.call(func, args = list(x = x, W = W, circular = FALSE))
  out.cir.T <- do.call(func, args = list(x = x, W = W, circular = TRUE))


  ## Set window of 2 plots
  # par(mfrow = c(2, 1))
  par(mfrow = c(3, 1))
  ## Predefine y axis ranges
  ylim.signalx <- range(x)
  ylim.signalx.max0 <- max(ylim.signalx) + 1
  ylim.signalx.min0 <- min(ylim.signalx) - 1
  ylim.out <- range(c(out.cir.F, out.cir.T), na.rm = TRUE)
  ylim.out.max0 <- max(ylim.out) + 1
  ylim.out.min0 <- min(ylim.out) - 1

  ## Sequence of x axis indices iteration
  seq.here <- sort(unique(c(seq(1, N - W + 1, by = by.val), N - W + 1)))
  ## Font size scalar
  # par.w2 <- 1
  par.w2 <- 1.6

  for(i in seq.here){

    # Set all plots size
    par(cex.lab = par.w2, cex.axis = par.w2 * 3/4, cex.main = par.w2, cex.sub = par.w2)

    ## Plot 1
    ##
    ## Margins area
    par(mar = c(4.1, 4.1, 4.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:N, rep(NA, N), type = 'l', xlab = '', ylab = '', ylim = ylim.signalx, xlim = c(0, N))
    ## Red shaded area
    polygon(c(i, i, (i+W-1), (i+W-1)),
            c(ylim.signalx.min0, ylim.signalx.max0, ylim.signalx.max0, ylim.signalx.min0),
            col=rgb(1, 0, 0, 0.05), border = NA )
    ## Plot signal x
    lines(1:N, x, type = 'l')
    ## Add title
    title(plt.title.vec[1], adj = 0, line = 1)
    ## Plot pattern border marks (vertical lines)
    abline(v = i, lty = 2, col = "red")
    abline(v = (i+W-1), lty = 2, col = "red")

    ## Plot 2
    ##
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.F[1:i], xlim = c(1, N), type = 'l', col = 'blue', lwd = 3,
         xlab = '', ylab = '',
         ylim = ylim.out)
    ## Add title
    title(plt.title.vec[2], adj = 0, line = 1)
    ## Add output border mark (vertical line)
    abline(v = i, lty = 2, col = "blue")

    ## Plot 3
    ##
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.F[1:i], xlim = c(1, N), type = 'l', col = 'blue', lwd = 3,
         xlab = 'Time [s]', ylab = '',
         ylim = ylim.out)
    ## Add title
    title(plt.title.vec[3], adj = 0, line = 1)
    ## Add output border mark (vertical line)
    abline(v = i, lty = 2, col = "blue")

    ## Sleep system before plotting the next plot
    if (!(i == max(seq.here))){
      Sys.sleep(0.1)
    } else {
      Sys.sleep(0.03)
    }

  }


  ## Set window of 3 plots
  par(mfrow = c(3, 1))
  ## Sequence of x axis indices iteration
  seq.here <- sort(unique(c(seq(N - W + 1, N,  by = 10), N)))
  ## Font size scalar
  par.w3 <- 1.6

  for(i in seq.here){

    ## If this is the last plot, clean the plot space
    if (i == max(seq.here)){
      dev.off()
      par(mfrow = c(3, 1))
    }

    # Set all plots size
    par(cex.lab = par.w3, cex.axis = par.w3* 2/3, cex.main = par.w3, cex.sub = par.w3)

    ## Plot 1
    ##
    ## Margins area
    par(mar = c(4.1, 4.1, 4.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:N, rep(NA, N), type = 'l', xlab = '', ylab = '', ylim = ylim.signalx, xlim = c(0,N))
    ## Red shaded area: part A
    polygon(c(i, i, N, N),
            c(ylim.signalx.min0, ylim.signalx.max0, ylim.signalx.max0, ylim.signalx.min0),
            col = rgb(1, 0, 0, 0.05), border = NA)
    ## Red shaded area: part B
    polygon(c(1, 1, (W - (N - i) - 1), (W - (N - i) - 1)),
            c(ylim.signalx.min0, ylim.signalx.max0, ylim.signalx.max0, ylim.signalx.min0),
            col=rgb(1, 0, 0, 0.05), border = NA )
    ## Plot signal x
    lines(1:N, x, type = 'l')
    ## Plot pattern: part A
    title(plt.title.vec[1], adj = 0, line = 1)
    abline(v = i, lty = 2, col = "red")
    ## Plot pattern: part B
    if (i == seq.here[1]){
      abline(v = N, lty = 2, col = "red")
    } else {
      pat.idx.B.end <- W - (N - i) - 1
      pat.idx.B <- 1:pat.idx.B.end
      abline(v = (pat.idx.B.end), lty = 2, col = "red")
    }


    ## Plot 2
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.F[1:i], xlim = c(0, N), type = 'l', col = 'blue', lwd = 3,
         xlab = '', ylab = '',
         ylim = range(c(out.cir.F, out.cir.T), na.rm = TRUE))
    title(plt.title.vec[2], adj = 0, line = 1)
    abline(v = i, lty = 2, col = "blue")

    ## Plot 3
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.T[1:i], xlim = c(0, N), type = 'l', col = 'blue', lwd = 3,
         xlab = 'Time [s]', ylab = '',
         ylim = range(c(out.cir.F, out.cir.T), na.rm = TRUE))
    title(plt.title.vec[3], adj = 0, line = 1)
    abline(v = i, lty = 2, col = "blue")

    ## Sleep system before plotting the next plot
    Sys.sleep(0.1)
  }

}




#' Generates plots for demo of package functions which take time series and
#' a short-timed pattern
#'
#' @param func runstats package core function
#' @param plt.title.vec vector of function-specific plot titles
#'
#' @importFrom grDevices rgb
#' @importFrom graphics abline lines par plot points polygon title
#'
#' @return \code{NULL}
#'
#' @examples
#' \dontrun{
#' func <- RunningCov
#' vec <- c("black: x\nred: running pattern",
#'          "RunningCov(x, pattern)",
#'          "RunningCov(x, pattern, circular = TRUE)")
#' plot.with.pattern(func, vec)
#' }
#'
#' @noRd
#'
plot.with.pattern <- function(func, plt.title.vec, func.name){

  ## Fixed constraints
  N <- 600
  x <- sin(seq(0, 2 * pi * 4, length.out = N))
  N.pat <- N / 3 + 1
  pat <- x[1:N.pat]
  pat.l <- length(pat)
  by.val <- 10

  ## Function output
  out.cir.F <- do.call(func, args = list(x = x, y = pat, circular = FALSE))
  out.cir.T <- do.call(func, args = list(x = x, y = pat, circular = TRUE))


  ## Set window of 2 plots
  # par(mfrow = c(2, 1))
  par(mfrow = c(3, 1))

  ## Predefine y axis ranges
  ylim.signalx <- c(-1,1)
  ylim.signalx.max0 <- max(ylim.signalx) + 1
  ylim.signalx.min0 <- min(ylim.signalx) - 1
  ylim.out <- range(c(out.cir.F, out.cir.T), na.rm = TRUE)
  ylim.out.max0 <- max(ylim.out) + 1
  ylim.out.min0 <- min(ylim.out) - 1

  ## Sequence of x axis indices iteration
  seq.here <- sort(unique(c(seq(1, N - N.pat + 1, by = by.val), N - N.pat + 1)))
  ## Font size scalar
  # par.w2 <- 1
  par.w2 <- 1.6

  for(i in seq.here){

    # Set all plots size
    par(cex.lab = par.w2, cex.axis = par.w2 * 3/4 , cex.main = par.w2, cex.sub = par.w2)

    ## Plot 1
    ##
    ## Margins area
    par(mar = c(4.1, 4.1, 4.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:N, rep(NA, N), type = 'l', xlab = '', ylab = '', ylim = ylim.signalx)
    ## Red shaded area
    polygon(c(i, i, (i+pat.l-1), (i+pat.l-1)),
            c(ylim.signalx.min0, ylim.signalx.max0, ylim.signalx.max0, ylim.signalx.min0),
            col=rgb(1, 0, 0, 0.05), border = NA )
    ## Plot signal x
    lines(1:N, x, type = 'l')
    ## Add title
    title(plt.title.vec[1], adj = 0, line = 1)
    ## Plot pattern
    lines(i:(i+pat.l-1), pat, col = rgb(1, 0, 0, alpha = 0.8), lwd = 3)
    ## Plot pattern border marks (vertical lines)
    abline(v = i, lty = 2, col = "red")
    abline(v = (i+pat.l-1), lty = 2, col = "red")

    ## Plot 2
    ##
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.F[1:i], xlim = c(0, N), type = 'l', col = 'blue', lwd = 3,
         xlab = '', ylab = '',
         ylim = ylim.out)
    ## Add title
    title(plt.title.vec[2], adj = 0, line = 1)
    ## Add output border mark (vertical line)
    abline(v = i, lty = 2, col = "blue")

    ## Plot 3
    ##
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.F[1:i], xlim = c(0, N), type = 'l', col = 'blue', lwd = 3,
         xlab = 'Time [s]', ylab = '',
         ylim = ylim.out)
    ## Add title
    title(plt.title.vec[3], adj = 0, line = 1)
    ## Add output border mark (vertical line)
    abline(v = i, lty = 2, col = "blue")

    ## Sleep system before plotting the next plot
    if (!(i == max(seq.here))){
      Sys.sleep(0.1)
    } else {
      Sys.sleep(0.03)
    }

  }


  ## Set window of 3 plots
  par(mfrow = c(3, 1))
  ## Sequence of x axis indices iteration
  seq.here <- sort(unique(c(seq(N - N.pat + 1, N,  by = 10), N)))
  ## Font size scalar
  par.w3 <- 1.6

  for(i in seq.here){

    ## If this is the last plot, clean the plot space
    if (i == max(seq.here)){
      dev.off()
      par(mfrow = c(3, 1))
    }

    # Set all plots size
    par(cex.lab = par.w3, cex.axis = par.w3 * 2/3, cex.main = par.w3, cex.sub = par.w3)

    ## Plot 1
    ##
    ## Margins area
    par(mar = c(4.1, 4.1, 4.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:N, rep(NA, N), type = 'l', xlab = '', ylab = '', ylim = ylim.signalx)
    ## Red shaded area: part A
    polygon(c(i, i, N, N),
            c(ylim.signalx.min0, ylim.signalx.max0, ylim.signalx.max0, ylim.signalx.min0),
            col = rgb(1, 0, 0, 0.05), border = NA)
    ## Red shaded area: part B
    polygon(c(1, 1, (N.pat - (N - i) - 1), (N.pat - (N - i) - 1)),
            c(ylim.signalx.min0, ylim.signalx.max0, ylim.signalx.max0, ylim.signalx.min0),
            col=rgb(1, 0, 0, 0.05), border = NA )
    ## Plot signal x
    lines(1:N, x, type = 'l')
    ## Plot pattern: part A
    title(plt.title.vec[1], adj = 0, line = 1)
    pat.idx.A <- i:N
    lines(pat.idx.A, pat[1:length(pat.idx.A)], col = rgb(1, 0, 0, alpha = 0.8), lwd = 3)
    if (length(pat.idx.A)==1){
      points(pat.idx.A, pat[1:length(pat.idx.A)], col = rgb(1, 0, 0, alpha = 0.8), lwd = 3, pch = 19)
    }
    abline(v = i, lty = 2, col = "red")
    ## Plot pattern: part B
    if (i == seq.here[1]){
      abline(v = N, lty = 2, col = "red")
    } else {
      pat.idx.B.end <- N.pat - (N - i) - 1
      pat.idx.B <- 1:pat.idx.B.end
      lines(pat.idx.B, pat[(N.pat - pat.idx.B.end + 1):N.pat], col = rgb(1, 0, 0, alpha = 0.8), lwd = 3)
      abline(v = (pat.idx.B.end), lty = 2, col = "red")
    }


    ## Plot 2
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.F[1:i], xlim = c(0, N), type = 'l', col = 'blue', lwd = 3,
         xlab = '', ylab = '',
         ylim = range(c(out.cir.F, out.cir.T), na.rm = TRUE))
    title(plt.title.vec[2], adj = 0, line = 1)
    abline(v = i, lty = 2, col = "blue")

    ## Plot 3
    ## Margins area
    par(mar = c(5.1, 4.1, 3.1, 2.1), mgp = c(3, 1, 0), las = 0) ## inside (b,l,t,r)
    ## Plot body
    plot(1:i, out.cir.T[1:i], xlim = c(0, N), type = 'l', col = 'blue', lwd = 3,
         xlab = 'Time [s]', ylab = '',
         ylim = range(c(out.cir.F, out.cir.T), na.rm = TRUE))
    title(plt.title.vec[3], adj = 0, line = 1)
    abline(v = i, lty = 2, col = "blue")

    ## Sleep system before plotting the next plot
    if (!(i == max(seq.here))){
      Sys.sleep(0.1)
    }
  }

}




#' Demo visualization of package functions
#'
#' Generates demo visualization of output of methods for computing running
#' statistics.
#'
#' @param func.name Character value; one of the following:
#' \itemize{
#'   \item "RunningMean",
#'   \item "RunningSd",
#'   \item "RunningVar",
#'   \item "RunningCov",
#'   \item "RunningCor",
#'   \item "RunningL2Norm".
#' }
#'
#' @return \code{NULL}
#'
#' @export
#'
#' @examples
#' \dontrun{
#' runstats.demo(func.name = "RunningMean")
#' runstats.demo(func.name = "RunningSd")
#' runstats.demo(func.name = "RunningVar")
#' runstats.demo(func.name = "RunningCov")
#' runstats.demo(func.name = "RunningCor")
#' runstats.demo(func.name = "RunningL2Norm")
#' }
#'
runstats.demo <- function(func.name = "RunningCov"){

  ## Match function name in argument with the package function
  func <- switch(func.name,
                 "RunningMean"   = RunningMean,
                 "RunningSd"     = RunningSd,
                 "RunningVar"    = RunningVar,
                 "RunningCov"    = RunningCov,
                 "RunningCor"    = RunningCor,
                 "RunningL2Norm" = RunningL2Norm)
  if (is.null(func)) stop("Undefined func.name argument value.")

  ## Match function name in argument with the plot title
  plt.title.vec <-  switch(func.name,
                           "RunningMean"   =  c("black: time-series x\nred: running window of length W",
                                               "RunningMean(x, W, circular = FALSE)",
                                               "RunningMean(x, W, circular = TRUE)"),

                           "RunningSd"     =  c("black: time-series x\nred: running window of length W",
                                               "RunningSd(x, W, circular = FALSE)",
                                               "RunningSd(x, W, circular = TRUE)"),

                           "RunningVar"    =  c("black: time-series x\nred: running window of length W",
                                               "RunningVar(x, W, circular = FALSE)",
                                               "RunningVar(x, W, circular = TRUE)"),

                           "RunningCov"    =  c("black: time-series x\nred: pattern y",
                                               "RunningCov(x, y, circular = FALSE)",
                                               "RunningCov(x, y, circular = TRUE)"),

                           "RunningCor"    =  c("black: time-series x\nred: pattern y",
                                               "RunningCor(x, y, circular = FALSE)",
                                               "RunningCor(x, y, circular = TRUE)"),

                           "RunningL2Norm" =  c("black: time-series x\nred: pattern y",
                                               "RunningL2Norm(x, y, circular = FALSE)",
                                               "RunningL2Norm(x, y, circular = TRUE)"))

  if (func.name %in% c("RunningCov", "RunningCor", "RunningL2Norm")){
    plot.with.pattern(func, plt.title.vec, func.name)
  } else {
    plot.no.pattern(func, plt.title.vec, func.name)
  }

}


vignettes/using-runstats.R
## ----setup, include = FALSE----------------------------------------------
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  cache = FALSE
)

## ---- eval = FALSE-------------------------------------------------------
#  devtools::install_github("martakarass/runstats")

## ---- eval = FALSE-------------------------------------------------------
#  library(runstats)
#  
#  ## Example: running correlation
#  x0 <- sin(seq(0, 2 * pi * 5, length.out = 1000))
#  x  <- x0 + rnorm(1000, sd = 0.1)
#  pattern <- x0[1:100]
#  out1 <- RunningCor(x, pattern)
#  out2 <- RunningCor(x, pattern, circular = TRUE)
#  
#  ## Example: running mean
#  x <- cumsum(rnorm(1000))
#  out1 <- RunningMean(x, W = 100)
#  out2 <- RunningMean(x, W = 100, circular = TRUE)

## ---- eval = FALSE-------------------------------------------------------
#  ## Example: demo for running correlation method
#  runstats.demo("RunningCor")

## ---- eval = FALSE-------------------------------------------------------
#  ## Example: demo for running mean method
#  runstats.demo("RunningMean")

## ---- echo=FALSE, include=FALSE------------------------------------------
library(rbenchmark)
library(ggplot2)

## ---- eval = FALSE-------------------------------------------------------
#  library(rbenchmark)
#  
#  set.seed (20181010)
#  x.N.seq <- 10^(3:7)
#  x.list  <- lapply(x.N.seq, function(N) runif(N))
#  y <- runif(100)
#  
#  ## Benchmark execution time of RunningCov
#  out.df <- data.frame()
#  for (x.tmp in x.list){
#    out.df.tmp <- benchmark("runstats" = runstats::RunningCov(x.tmp, y),
#                            replications = 10,
#                            columns = c("test", "replications", "elapsed",
#                                        "relative", "user.self", "sys.self"))
#    out.df.tmp$x_length <- length(x.tmp)
#    out.df.tmp$pattern_length <- length(y)
#    out.df <- rbind(out.df, out.df.tmp)
#  }
#  

## ---- echo = FALSE, eval = FALSE-----------------------------------------
#  
#  out.df.path <- "../inst/benchmark_results/2018-12-17-performance_rbenchmark_suite1.csv"
#  write.csv(out.df, out.df.path, row.names = FALSE, quote = FALSE)

## ---- echo = FALSE, eval = TRUE------------------------------------------
out.df.path <- "../inst/benchmark_results/2018-12-17-performance_rbenchmark_suite1.csv"
out.df <- read.csv(out.df.path)
options("scipen" = 10, "digits" = 4)

## ------------------------------------------------------------------------
knitr::kable(out.df)

## ---- eval = FALSE-------------------------------------------------------
#  ## Conventional approach
#  RunningCov.sapply <- function(x, y){
#    l_x <- length(x)
#    l_y <- length(y)
#    sapply(1:(l_x - l_y + 1), function(i){
#      cov(x[i:(i+l_y-1)], y)
#    })
#  }
#  
#  set.seed (20181010)
#  
#  out.df2 <- data.frame()
#  for (x.tmp in x.list[c(1,2,3,4)]){
#    out.df.tmp <- benchmark("conventional" = RunningCov.sapply(x.tmp, y),
#                            "runstats" = runstats::RunningCov(x.tmp, y),
#                            replications = 10,
#                            columns = c("test", "replications", "elapsed",
#                                        "relative", "user.self", "sys.self"))
#    out.df.tmp$x_length <- length(x.tmp)
#    out.df2 <- rbind(out.df2, out.df.tmp)
#  }
#  

## ---- echo = FALSE, eval = FALSE-----------------------------------------
#  out.df.path <-"../inst/benchmark_results/2018-12-17-performance_rbenchmark_suite2.csv"
#  write.csv(out.df2, out.df.path, row.names = FALSE, quote = FALSE)

## ---- echo = FALSE, eval = TRUE------------------------------------------
out.df.path <- "../inst/benchmark_results/2018-12-17-performance_rbenchmark_suite2.csv"
out.df2 <- read.csv(out.df.path)

## ---- fig.width=10, fig.height=5-----------------------------------------
library(ggplot2)

plt1 <- 
  ggplot(out.df2, aes(x = x_length, y = elapsed, color = test)) + 
  geom_line() + geom_point(size = 3) + scale_x_log10() + 
  theme_minimal(base_size = 14) + 
  labs(x = "Vector length of x",
       y = "elapsed [s]", color = "Method", 
       title = "Running covariance rbenchmark") + 
  theme(legend.position = "bottom")
plt2 <- 
  plt1 + 
  scale_y_log10() + 
  labs(y = "log of elapsed [s]")

cowplot::plot_grid(plt1, plt2, nrow = 1, labels = c('A', 'B'))

## ------------------------------------------------------------------------
sessioninfo::platform_info()

